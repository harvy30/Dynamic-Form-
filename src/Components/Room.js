import React from 'react';
import styled from 'styled-components';

const RoomForm = styled.form`
    background-color:#D3D3D3 !important;
`;

const RoomBody = styled.div`
    ${({bodyClass}) => {
      return  bodyClass && `
         background-color:#fff !important;
    `
    }}
`;

const Room = (props) => {

  return (
    <RoomForm className=" card p-2 m-3" disabled>
      <div className="room-header text-left mb-3 font-weight-bold">
        {!props.hideCheckbox && (
          <input type="checkbox"
            id={`room-chk-${props.index}`}
            className="mx-2 room-check-box"
            checked={props.checked}
            onChange={props.onHandleChange(props.RoomNo)}
            value={props.RoomNo}
          />)}
        <label htmlFor={`room-chk-${props.index}`}>{props.title}</label>
      </div>
      <fieldset disabled={props.disabled}>

        <RoomBody className="p-3" bodyClass={props.bodyClass}>
          <div className="d-flex justify-content-between">
            <label className="mx-3" htmlFor="adults">Adults <br />(18+)</label>
            <label className="mx-3" htmlFor="children">Children<br /> (0-17)</label>
          </div>
          <div className="d-flex justify-content-between">
            <select
              id="adults"
              className="mx-3"
              value={props.Adults.find(a => a == props.adult)}
              onChange={props.onHandleDropDown('adult', props.RoomNo)}>
              {props.Adults.map(a => {
                return (<option key={a} value={a}>{a}</option>)
              })}
            </select>

            <select
              id="children"
              className="mx-3"
              value={props.Children.find(a => a == props.child)}
              onChange={props.onHandleDropDown('child', props.RoomNo)}>
              {props.Children.map(c => {
                return (<option key={c} value={c}>{c}</option>)
              })}
            </select>

          </div>
        </RoomBody>
      </fieldset>
    </RoomForm>
  )
}

export default Room;
