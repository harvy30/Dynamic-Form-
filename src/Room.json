[
    {
        "title":"Room 1",
        "checked":true,
        "RoomNo":"1",
        "Adults":[1,2],
        "Children":[0,1,2],
        "disabled": false,
        "hideCheckbox":true,
        "bodyClass":"room-body"
    },
    {
    "title":"Room 2",
    "checked":false,
    "RoomNo":"2",
    "Adults":[1,2],
    "Children":[0,1,2],
    "disabled": true
    },
    {
        "title":"Room 3",
        "checked":false,
        "RoomNo":"3",
        "Adults":[1,2],
        "Children":[0,1,2],
        "disabled": true
    },
    {
        "title":"Room 4",
        "checked":false,
        "RoomNo":"4",
        "Adults":[1,2],
        "Children":[0,1,2],
        "disabled": true
    }
 
]