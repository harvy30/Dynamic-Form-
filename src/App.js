import React from 'react';
import Room from './Components/Room'
import RoomData from './Room.json'
import 'bootstrap/dist/css/bootstrap.css';
import './App.css';

function App() {
  const roomDataFinal = localStorage.getItem('room') ? JSON.parse(localStorage.getItem('room')) : RoomData;
  const [roomData, setRoomData] = React.useState([...roomDataFinal])

  const onSubmit = () => {
    localStorage.setItem("room", JSON.stringify(roomData));
    alert("Room Saved Successfully!!!");
  }

  const onHandleDropDown = (name, rNo) => event => {
    const roomIndex = roomData.findIndex(r => r.RoomNo === rNo);
    roomData[roomIndex] = { ...roomData[roomIndex], [name]: event.target.value };
    setRoomData([...roomData]);
  };

  const getRoom = ({ checked, r }) => {
    return {
      ...r,
      checked: checked,
      bodyClass: checked ? "room-body" : "",
      disabled: !checked,
      adult: !checked ? "0" : r.adult,
      child: !checked ? "0" : r.child
    }
  }

  const onHandleChange = rNo => event => {
    const rooms = roomData.map((r, idx, rooms) => {
      if (r.RoomNo >= rNo && !r.hideCheckbox && !event.target.checked) {
        rooms[idx] = getRoom({ checked: event.target.checked, r });
      }
      if (r.RoomNo <= rNo && !r.hideCheckbox && event.target.checked) {
        rooms[idx] = getRoom({ checked: event.target.checked, r });
      }
      return rooms[idx];
    });
    setRoomData([...rooms]);
  }

  return (
    <div className="App">
      <div className="d-flex" autoComplete="off">
        {roomData.map((room, idx) => {
          return <Room {...room} key={room.RoomNo}
            onHandleChange={onHandleChange}
            onHandleDropDown={onHandleDropDown} />
        })}

      </div>
      <div className="float-left mx-5">
        <input id="btnSubmit" type="button" value="Submit" onClick={() => onSubmit()} />
      </div>

    </div>
  );
}

export default App;
