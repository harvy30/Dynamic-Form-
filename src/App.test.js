import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import { mount, shallow } from 'enzyme'
import Room from './Components/Room'
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import RoomData from './Room.json'

configure({ adapter: new Adapter() });


it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});


it('it should render 4 Rooms', () => {
  const wrapper = mount(<App />);
  const roomComponents = wrapper.find("Room");
  expect(roomComponents.length).toEqual(4);
});


it('should render correctly', () => {
  const output = renderer
    .create(<App />)
    .toJSON();
  expect(output).toMatchSnapshot();
});

it('it should render submit Button', () => {
  const wrapper = mount(<App />);
  const wrapperInstance = shallow(<App />);
  expect(wrapperInstance).toEqual({});

  const btnSubmit = wrapper.find("#btnSubmit");
  expect(btnSubmit.length).toEqual(1);
});



it('it should render Rooms component', () => {
  const wrapper = mount(<App />);
  const onHandleChange = jest.fn();
  const onHandleDropDown = jest.fn();

  const roomComponents = wrapper.find(Room);

  const instancesRoom = RoomData.map((room, idx) => shallow(<Room {...room} key={room.RoomNo}
    onHandleChange={onHandleChange}
    onHandleDropDown={onHandleDropDown}
  />));

  roomComponents.forEach((room, i) => {
    const chkBox = room.find('.room-check-box');
    const props = room.props();
    if (i === 0) {
      expect(chkBox.length).toEqual(0)
    } else {
      chkBox.simulate('change');
      expect(onHandleChange).toHaveBeenCalledWith(props.RoomNo)
    }
  })
  expect(roomComponents.length).toEqual(instancesRoom.length);
});


