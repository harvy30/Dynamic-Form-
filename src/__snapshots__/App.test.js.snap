// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="App"
>
  <div
    autoComplete="off"
    className="d-flex"
  >
    <form
      className="room card p-2 m-3"
      disabled={true}
    >
      <div
        className="room-header text-left mb-3 font-weight-bold"
      >
        Room 1
      </div>
      <fieldset
        disabled={false}
      >
        <div
          className="p-3 room-body"
        >
          <div
            className="d-flex justify-content-between"
          >
            <label
              className="mx-3"
              htmlFor="age-simple"
            >
              Adults 
              <br />
              (18+)
            </label>
            <label
              className="mx-3"
              htmlFor="age-simple"
            >
              Children
              <br />
               (0-17)
            </label>
          </div>
          <div
            className="d-flex justify-content-between"
          >
            <select
              className="mx-3"
              onChange={[Function]}
            >
              <option
                value={1}
              >
                1
              </option>
              <option
                value={2}
              >
                2
              </option>
            </select>
            <select
              className="mx-3"
              onChange={[Function]}
            >
              <option
                value={0}
              >
                0
              </option>
              <option
                value={1}
              >
                1
              </option>
              <option
                value={2}
              >
                2
              </option>
            </select>
          </div>
        </div>
      </fieldset>
    </form>
    <form
      className="room card p-2 m-3"
      disabled={true}
    >
      <div
        className="room-header text-left mb-3 font-weight-bold"
      >
        <input
          checked={false}
          className="mx-2 room-check-box"
          onChange={[Function]}
          type="checkbox"
          value="2"
        />
        Room 2
      </div>
      <fieldset
        disabled={true}
      >
        <div
          className="p-3 undefined"
        >
          <div
            className="d-flex justify-content-between"
          >
            <label
              className="mx-3"
              htmlFor="age-simple"
            >
              Adults 
              <br />
              (18+)
            </label>
            <label
              className="mx-3"
              htmlFor="age-simple"
            >
              Children
              <br />
               (0-17)
            </label>
          </div>
          <div
            className="d-flex justify-content-between"
          >
            <select
              className="mx-3"
              onChange={[Function]}
            >
              <option
                value={1}
              >
                1
              </option>
              <option
                value={2}
              >
                2
              </option>
            </select>
            <select
              className="mx-3"
              onChange={[Function]}
            >
              <option
                value={0}
              >
                0
              </option>
              <option
                value={1}
              >
                1
              </option>
              <option
                value={2}
              >
                2
              </option>
            </select>
          </div>
        </div>
      </fieldset>
    </form>
    <form
      className="room card p-2 m-3"
      disabled={true}
    >
      <div
        className="room-header text-left mb-3 font-weight-bold"
      >
        <input
          checked={false}
          className="mx-2 room-check-box"
          onChange={[Function]}
          type="checkbox"
          value="3"
        />
        Room 3
      </div>
      <fieldset
        disabled={true}
      >
        <div
          className="p-3 undefined"
        >
          <div
            className="d-flex justify-content-between"
          >
            <label
              className="mx-3"
              htmlFor="age-simple"
            >
              Adults 
              <br />
              (18+)
            </label>
            <label
              className="mx-3"
              htmlFor="age-simple"
            >
              Children
              <br />
               (0-17)
            </label>
          </div>
          <div
            className="d-flex justify-content-between"
          >
            <select
              className="mx-3"
              onChange={[Function]}
            >
              <option
                value={1}
              >
                1
              </option>
              <option
                value={2}
              >
                2
              </option>
            </select>
            <select
              className="mx-3"
              onChange={[Function]}
            >
              <option
                value={0}
              >
                0
              </option>
              <option
                value={1}
              >
                1
              </option>
              <option
                value={2}
              >
                2
              </option>
            </select>
          </div>
        </div>
      </fieldset>
    </form>
    <form
      className="room card p-2 m-3"
      disabled={true}
    >
      <div
        className="room-header text-left mb-3 font-weight-bold"
      >
        <input
          checked={false}
          className="mx-2 room-check-box"
          onChange={[Function]}
          type="checkbox"
          value="4"
        />
        Room 4
      </div>
      <fieldset
        disabled={true}
      >
        <div
          className="p-3 undefined"
        >
          <div
            className="d-flex justify-content-between"
          >
            <label
              className="mx-3"
              htmlFor="age-simple"
            >
              Adults 
              <br />
              (18+)
            </label>
            <label
              className="mx-3"
              htmlFor="age-simple"
            >
              Children
              <br />
               (0-17)
            </label>
          </div>
          <div
            className="d-flex justify-content-between"
          >
            <select
              className="mx-3"
              onChange={[Function]}
            >
              <option
                value={1}
              >
                1
              </option>
              <option
                value={2}
              >
                2
              </option>
            </select>
            <select
              className="mx-3"
              onChange={[Function]}
            >
              <option
                value={0}
              >
                0
              </option>
              <option
                value={1}
              >
                1
              </option>
              <option
                value={2}
              >
                2
              </option>
            </select>
          </div>
        </div>
      </fieldset>
    </form>
  </div>
  <div
    className="float-left mx-5"
  >
    <input
      id="btnSubmit"
      onClick={[Function]}
      type="button"
      value="Submit"
    />
  </div>
</div>
`;
